src/peft/tuners/tuners_utils.py

```diff
@@ -523,3 +528,41 @@ def inspect_matched_modules(tuner: BaseTuner, adapter_name: str = "default") ->
         else:
             module_dict["unmatched"].append(key)
     return module_dict
+
+
+def _maybe_include_all_linear_layers(peft_config: PeftConfig, model: nn.Module) -> PeftConfig:
+    """
+    Helper function to update `target_modules` to all linear/Conv1D layers if provided as 'all_linear'. Adapted from
+    the QLoRA repository: https://github.com/artidoro/qlora/blob/main/qlora.py
+    """
+    # if `target_modules` is a string, forgive case issues and match
+    if (
+        isinstance(peft_config.target_modules, str)
+        and peft_config.target_modules.lower() == INCLUDE_LINEAR_LAYERS_SHORTHAND
+    ):
+        is_loaded_in_8bit = getattr(model, "is_loaded_in_8bit", False)
+        is_loaded_in_4bit = getattr(model, "is_loaded_in_4bit", False)
+        # match with a list of linear layer classes. this is needed as sometimes you can
+        # have a mix Eg. T5 with 8bit has instances of torch.nn.Linear and bnb.nn.Linear8bitLt
+        linear_classes = (torch.nn.Linear, Conv1D)
+        if is_bnb_available():
+            import bitsandbytes as bnb
+
+            linear_classes = (bnb.nn.Linear8bitLt,) + linear_classes if is_loaded_in_8bit else linear_classes
+            linear_classes = (bnb.nn.Linear4bit,) + linear_classes if is_loaded_in_4bit else linear_classes
+
+        linear_module_names = set()
+        for name, module in model.named_modules():
+            # match with all linear classes.
+            if isinstance(module, linear_classes):
+                names = name.split(".")
+                linear_module_names.add(names[-1])  # get the base name
+
+        # ignore the last classification head for text generation models
+        output_emb = model.get_output_embeddings()
```

Comment:
You're right! However, on closer inspection, it looks like this function makes sense only for `PreTrainedModel` subclasses - firstly, because this is meant to be simple shorthand to mimic QLoRa-like training. Further, the more complicated `ModuleList` structures in `diffusers` models makes the matching function incorrect. In these cases, I'll need to save the full name of the module, instead of the base name, but then I'm also trying to save everything into the PeftConfig, so now the `target_modules` list is very long (can be 1000s) for LLMs. It made more sense for me to raise a ValueError in case the model is not a PreTrainedModel. I've added this now and have also added a test case for a diffusion model where this error is raised. 

This function now doesn't work for simple `nn.Module` models, but then it's trivial to get the same `all-linear` functionality for such models. So I think this should be fine!

LMK what you think!